modules/datastore/modules/datastore_mysql_import/src/Service/MysqlImport.php

```diff
@@ -61,26 +71,74 @@ private function getDatabaseConnectionCapableOfDataLoad() {
   }
 
   /**
-   * Private.
+   * Properly escape and format the supplied list of column names.
+   *
+   * @param string[] $columns
+   *   List of column names.
+   *
+   * @return array
+   *   List of sanitized table headers keyed by original column names.
    */
-  private function cleanHeaders($headers) {
-    $row = [];
-    foreach ($headers as $field) {
-      $new = preg_replace("/[^A-Za-z0-9_ ]/", '', $field);
-      $new = trim($new);
-      $new = strtolower($new);
-      $new = str_replace(" ", "_", $new);
-
-      $mysqlMaxColLength = 64;
-      if (strlen($new) > $mysqlMaxColLength) {
-        $strings = str_split($new, $mysqlMaxColLength - 5);
-        $token = substr(md5($field), 0, 4);
-        $new = $strings[0] . "_{$token}";
-      }
-
-      $row[] = $new;
+  private function generateTableHeaders(array $columns): array {
+    return array_replace([], ...array_map(function ($column) {
+      // Sanitize the supplied table header to generate a unique column name.
+      $header = $this->sanitizeHeader($column);
+      // Prepend "d_" to the table column name to prevent numeric column name
+      // parsing issues.
+      //
+      // @see https://github.com/GetDKAN/dkan/issues/3606
+      $header = 'd_' . $header;
+      // Truncate the generated table column name, if necessary, to fit the max
+      // column length.
+      $header = $this->truncateHeader($header);
+
+      return [$column => $header];
+    }, $columns));
+  }
+
+  /**
+   * Sanitize table column name according to the MySQL supported characters.
+   *
+   * @param string $column
+   *   The column name being sanitized.
+   *
+   * @returns string
+   *   Sanitized column name.
+   */
+  protected function sanitizeHeader(string $column): string {
+    // Replace all spaces with underscores since spaces are not a supported
+    // character.
+    $column = str_replace(' ', '_', $column);
+    // Strip unsupported characters from the header.
+    $column = preg_replace('/[^A-Za-z0-9_ ]/', '', $column);
+    // Trim underscores from the beginning and end of the column name.
+    $column = trim($column, '_');
+    // Convert the column name to lowercase.
+    $column = strtolower($column);
+
+    return $column;
+  }
+
+  /**
+   * Truncate column name if longer than the max column length for the database.
+   *
+   * @param string $column
+   *   The column name being truncated.
+   *
+   * @returns string
+   *   Truncated column name.
+   */
+  protected function truncateHeader(string $column): string {
+    // If the supplied table column name is longer than the max column length,
+    // truncate the column name to 5 characters under the max length and
+    // substitute the truncated characters with a unique hash.
+    if (strlen($column) > self::MAX_COLUMN_LENGTH) {
+      $field = substr($column, 0, self::MAX_COLUMN_LENGTH - 5);
+      $hash = substr(md5($column), 0, 4);
```

Comment:
Similar question to the d_ prefix here - should this be made transparent to users?

Reply:
Not completely up to date on this issue, but in general, users want control over their columns and column names. If anything comes out different than it went in, people are going to be unhappy.

Reply:
This particular line pre-existed this PR. It's a constraint of MySQL. MySQL column names can be no longer than 64 characters and this appears to solve that constraint by replacing the last 4 characters with a segment from an md5 hash generated using the whole column name.
Regardless, this PR in conjunction with https://github.com/GetDKAN/datastore/pull/35 solves this issue by using the actual CSV column name provided as the preview column name for datasets while only using the truncated `d_` prefixed column name in MySQL.
TL;DR: the user shouldn't encounter a modified column name with this approach.